### ISSUE TITLE
Using gographics to convert RGB PNGs to 8 bit colormaps concurrently in goroutines

### ISSUE THREAD

**Issue by nw31304** [gographics/imagick#60]
Hiya.  Attempting to generate indexed (that is, palletized to 8 bit colormap) PNGs representing maptiles.  I generate numerous goroutines, and the logic in a go routine essentially is as follows:

1) Invoke MapNik to generate a 4096x4096 RGB PNG "supertile"
2) Use the Go Image package to carve up the supertile into 256 256x256 pixel PNG maptiles
3) (Sequentially) For each of the 256 tiles, call MagigickWand.Clear(), then invoke MagickWand's Quantize() function to create an indexed PNG using 256 colors
4) Write each palletized PNG to a file via MagickWand's WriteImageFile.

Having various issues -- it almost appears that IM / MW is not threadsafe, although the developers state otherwise.  Here's a typical SEGFAULT:

tiler: magick/colorspace.c:1528: TransformImageColorspace: Assertion `image->signature == 0xabacadabUL' failed.
SIGABRT: abort
PC=0x7f0608a47cc9 m=10
signal arrived during cgo execution

goroutine 45 [syscall, locked to thread]:
runtime.cgocall(0x4213e0, 0xc820463b08, 0x0)
	/usr/local/go/src/runtime/cgocall.go:120 +0x11b fp=0xc820463ac0 sp=0xc820463a90
github.com/gographics/imagick/imagick._Cfunc_MagickSetImageType(0x7f05e40015d0, 0x4, 0x7f0500000000)
	??:0 +0x39 fp=0xc820463b08 sp=0xc820463ac0
github.com/gographics/imagick/imagick.(*MagickWand).SetImageType(0xc82014a008, 0x4, 0x0, 0x0)
	/root/work/src/github.com/gographics/imagick/imagick/magick_wand_image.go:2609 +0x3b fp=0xc820463b28 sp=0xc820463b08
main.createTile(0xc8205d4990, 0x2f, 0x7f05f1a2f700, 0xc82016f240, 0x7f05f1a2f738, 0xc820448000, 0xc820463d40, 0xc82044a000, 0xc82014a008, 0x0, ...)
	/root/work/src/tomtomapac.com/tiler/tiler.go:250 +0x3b1 fp=0xc820463ca8 sp=0xc820463b28
main.tileWorker(0xc8200142a0, 0xc820014300, 0xc8200141e0, 0xc82014a008)
	/root/work/src/tomtomapac.com/tiler/tiler.go:340 +0x9dc fp=0xc820463f80 sp=0xc820463ca8
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1696 +0x1 fp=0xc820463f88 sp=0xc820463f80
created by main.main
	/root/work/src/tomtomapac.com/tiler/tiler.go:472 +0xaf1

I've tried calling imagick.Initialize() in the mainthread before the goroutines are spun up as well as having each goroutine call imagick.Initialize() locally.  I've tried creating a single MagickWand instance in the mainthread and sharing it among all goroutines, and I've tried having each goroutine create its own MagicWand instance.  In all cases, the above and similar errors occur under heavy load.

I'm running this under Ubuntu, using version 6.7.7-10 2014-03-06 Q16 of ImageMagick, and libMagickWand.so.5.0.0 is the lib the go binary is linked against..  I realize I've left out a lot of information you may need, but feel free to flame me back and state whatever further detail you require.  Essentially, it would really help if you would give me an idea how to structure my interaction with MagicWand in realtion to my goroutines.  Any other gotchas I need to be aware of when running MagickWand in concurrent goroutines?

**Comment by justinfx** [comment]
I can say that I am pretty confident that a MagickWand instance is not
thread safe and should not be shared. I have a similar application where
many http requests would come in and need image conversions to take place.
In my application, I would initialize imagick once from the application
main, and then I had a pool of 4 image workers that would receive requests
on a channel, and create and destroy a new wand each time to process an
image. Image processing would involve reading a source image from disk,
resizing, and then taking the image blob data and serving it. Hadn't seen
an issue with this setup.
How many concurrent MagickWands is your application using? Have you tried
reducing the number to see if it makes a difference? It's possible that
because I have really only tried up to about 8-10 image workers while
testing that I might not have brought about the crash. Or maybe it is the
difference in the parts of the API we are using.
Are you able to create a small reproduction of the issue that I could run?
Have you tried creating and destroying a new wand for each image operating
instead of sharing or resuing ?

**Comment by nw31304** [comment]
I've managed to isolate the issue to the small goroutine that actually writes the PNG images to a filename.  The ImageMagick is done in the main thread before the 8 goroutines are started.  As you can see, the goroutine is passed from a higher level routine the following parms:

fileName: name of the file to write the 256x256 indexed PNG to
srcImage:  a 4096x4096 NRGBA supertile created by MapNik
dstImage: a 256x256 NRGBA image
TileRect: 256x256 rectangle definition
srcPoint: location in the SuperTile from which to begin to carve the 256x256 sub-tile

So the first thing is to draw the sub-tile into dstImage, encode the 256x256 NRGBA image into a Buffer in PNG format, then create a new MagickWand and load the Blob into it.  Next, set a few informational items into the Wand's Image and Quantize() it.Next, check if this is an empty ocean tile, and if so, go no further (the web server will return a default ocean tile instead of a 404).  Now, here's where things get interesting.  If I open the file with os.OpenFile and use mw.WriteImageFile(), all hell breaks loose.  I get numerous errors of this form: 2015/10/28 01:40:23 Unable to encode png into file: /root/work/tiles/16/2367488/2367488_5804032.png;  Source point: (x,y) = (512,1792); err: bad file descriptor.  There is nothting wrong with the file descriptor, thank you.  However, if I replace the call with a dumb nw.WriteImage(fileName), everything works fine.  IOW, passing an open file seems to cause some sort of issue within IM that I can't quite track down, but passing the fileName itself makes it happy:

Here's the code:

```go
func createTile(fileName string, srcImage image.Image, dstImage draw.Image, srcPoint *image.Point) error {

var b bytes.Buffer

     draw.Draw(dstImage, TileRect, srcImage, *srcPoint, draw.Src)
 
     err := png.Encode(&b, dstImage)
     if err != nil {
         fmt.Println("error from png.Encode: ", err)
         return err
      }
      mw := imagick.NewMagickWand()
      defer mw.Destroy()
      mw.ReadImageBlob(b.Bytes())
      mw.SetImageType(imagick.IMAGE_TYPE_PALETTE)
      mw.SetImageCompression(imagick.COMPRESSION_ZIP)
      mw.SetImageFilename(fileName)
      err = mw.QuantizeImage(256, imagick.COLORSPACE_RGB, 0, false, false)
      if err != nil {
          fmt.Println("error from imagick.Quantitize: ", err)
          return err
      }
  
      if mw.GetImageColors() == 1 {
          pw, err := mw.GetImageColormapColor(0)
          if err != nil {
              fmt.Println("error from mw.GetImageColormapColor: ", err)
              return err
          }
          defer pw.Destroy()
          if pw.GetColorAsString() == ocean_tile {
              return nil
          }
      }
  
      outFile, err := os.OpenFile(fileName, os.O_WRONLY|os.O_CREATE, 0755)
  
      if err != nil && os.IsNotExist(err) {
          err = os.MkdirAll(path.Dir(fileName), os.FileMode(os.ModeDir|0755))
          if err != nil {
              return err
          }
          outFile, err = os.Create(fileName)
          if err != nil {
              return err
          }
      }
  
      defer outFile.Close()
  
      err = mw.WriteImageFile(outFile)
      //err = mw.WriteImage(fileName)
  
      if err != nil {
           return fmt.Errorf("Unable to encode png into file: %s;  Source point: (x,y) = (%d,%d); err: %s", fileName, srcPoint.X, srcPoint.Y, err)
       }
   
       return nil
   }
```

**Comment by nw31304** [comment]
Here's a test program that seems to elicit the same "bad file descriptor" behaviour:

```go
import (
	"bytes"
	"fmt"
	"github.com/gographics/imagick/imagick"
	"image"
	"image/png"
	"os"
	"time"
)

func writePng(fn string, count int, sleep_ms int, c chan<- int) {
	var err error

	for i := 0; i < count; i++ {
		// Create a blank image 256x256 pixels
		myImage := image.NewRGBA(image.Rect(0, 0, 256, 256))

		var b bytes.Buffer

		err = png.Encode(&b, myImage)
		if err != nil {
			fmt.Println("error from png.Encode: ", err)
			continue
		}

		mw := imagick.NewMagickWand()
		defer mw.Destroy()

		mw.ReadImageBlob(b.Bytes())
		mw.SetImageType(imagick.IMAGE_TYPE_PALETTE)
		mw.SetImageCompression(imagick.COMPRESSION_ZIP)
		err = mw.QuantizeImage(256, imagick.COLORSPACE_RGB, 0, false, false)
		if err != nil {
			fmt.Println("error from imagick.Quantitize: ", err)
			return
		}

		outFile, err := os.OpenFile(fn, os.O_WRONLY|os.O_CREATE, 0755)
		if err != nil {
			fmt.Printf("%s:%v\n", "Unable to open output file", err)
			continue
		}

		err = mw.WriteImageFile(outFile)
		if err != nil {
			fmt.Printf("%s:%v\n", "Unable to encode png file", err)
		}
		outFile.Close()
		time.Sleep(time.Duration(sleep_ms) * time.Millisecond)
	}
	c <- 1
}

func main() {
	imagick.Initialize()
	defer imagick.Terminate()

	c := make(chan int)

	for i := 0; i < 8; i++ {
		go writePng(fmt.Sprintf("/tmp/r%d.png", i), 100, 50, c)
	}

	for i := 0; i < 8; i++ {
		<-c
	}
}
```

**Comment by nw31304** [comment]
Here's a small sample program that seems to demonstrate the problem I'm having.  The main routine initializes IM, defers a Terminate, then spins up 8 goroutines that encode a PNG image to a bute buffer, quantitize it, opens an putput file, then writes the wand's image to the output file using the already opened file as a target.  On my machine, I get several "bad file decriptor" messages, followed by a stack trace.

```go
import (
	"bytes"
	"fmt"
	"github.com/gographics/imagick/imagick"
	"image"
	"image/png"
	"os"
	"time"
)

func writePng(fn string, count int, sleep_ms int, c chan<- int) {
	var err error

	for i := 0; i < count; i++ {
		// Create a blank image 256x256 pixels
		myImage := image.NewRGBA(image.Rect(0, 0, 256, 256))

		var b bytes.Buffer

		err = png.Encode(&b, myImage)
		if err != nil {
			fmt.Println("error from png.Encode: ", err)
			continue
		}

		mw := imagick.NewMagickWand()
		defer mw.Destroy()

		mw.ReadImageBlob(b.Bytes())
		mw.SetImageType(imagick.IMAGE_TYPE_PALETTE)
		mw.SetImageCompression(imagick.COMPRESSION_ZIP)
		err = mw.QuantizeImage(256, imagick.COLORSPACE_RGB, 0, false, false)
		if err != nil {
			fmt.Println("error from imagick.Quantitize: ", err)
			return
		}

		outFile, err := os.OpenFile(fn, os.O_WRONLY|os.O_CREATE, 0755)
		if err != nil {
			fmt.Printf("%s:%v\n", "Unable to open output file", err)
			continue
		}

		err = mw.WriteImageFile(outFile)
		if err != nil {
			fmt.Printf("%s:%v\n", "Unable to encode png file", err)
		}
		outFile.Close()
		time.Sleep(time.Duration(sleep_ms) * time.Millisecond)
	}
	c <- 1
}

func main() {
	imagick.Initialize()
	defer imagick.Terminate()

	c := make(chan int)

	for i := 0; i < 8; i++ {
		go writePng(fmt.Sprintf("/tmp/r%d.png", i), 100, 50, c)
	}

	for i := 0; i < 8; i++ {
		<-c
	}
}

```

**Comment by justinfx** [comment]
I am going to give this sample a try tomorrow. But, is there a bug in your code,  causing it to hammer the same output filepath in all goroutines? 

```go
for i := 0; i < worker_count; i++ {
     // not making sure to use a local scoped counter? 
    // i := i
    go writePng(fmt.Sprintf("/tmp/r%d.png", i), num_iterations, sleep_ms, c)
        }
```

**Comment by nw31304** [comment]
Nah - -  we're cool,  not an inner func or func literal to close over `i` .  Each goroutine gets it's own file. Always wind up with `num_workers` files in the /top dir.  BTW,  you guys put a ton of work into this project. Much appreciated.

**Comment by nw31304** [comment]
Even simpler:

```go
package main

import (
        "bytes"
        "fmt"
        "github.com/gographics/imagick/imagick"
        "image"
        "image/png"
        "os"
)

func main() {
        imagick.Initialize()
        defer imagick.Terminate()

        myImage := image.NewRGBA(image.Rect(0, 0, 256, 256))
        var err error
        var b bytes.Buffer

        err = png.Encode(&b, myImage)
        if err != nil {
                fmt.Println("error from png.Encode: ", err)
                return
        }

        mw := imagick.NewMagickWand()
        defer mw.Destroy()

        mw.ReadImageBlob(b.Bytes())

        out1, err := os.OpenFile("/tmp/file1", os.O_WRONLY|os.O_CREATE, 0755)
        if err != nil {
                fmt.Printf("%s:%v\n", "Unable to open output file", err)
                return
        }
        defer out1.Close()

        err = mw.WriteImageFile(out1)

        if err != nil {
                fmt.Printf("Unable to encode png file: %v\n", err)
        }

        out2, err := os.OpenFile("/tmp/file2", os.O_WRONLY|os.O_CREATE, 0755)
        if err != nil {
                fmt.Printf("%s:%v\n", "Unable to open output file", err)
                return
        }
        defer out2.Close()
        fmt.Println("out1 fd: ", out1.Fd())
        fmt.Println("out2 fd: ", out2.Fd())
}

```

I get this output:

```
out1 fd:  3
out2 fd:  3
```

Oh yeah, I'm sure this is gonna end REAL well...

**Comment by justinfx** [comment]
I gave your examples a test and see what you mean. The underlying Imagick code closes the file descriptor while the calling code still has the `os.File`. I've added a dup to the `cfdopen()`:
https://github.com/gographics/imagick/commit/5c118bd2d4ac87590d6927b500a9767e5f2fab7c

Can you give the branch `60_bad_file_desc` a test?

**Comment by nw31304** [comment]
Works perfectly -- well done.  Thanks for your quick response.

**Comment by justinfx** [comment]
Sure, no problem. Thanks for the really detailed reproductions! It make it a lot easier to track down and fix.

**Comment by justinfx** [comment]
Merged to master and im-6.8.9
